File: src/main/java/team404/PlayerRevivalCache.java
package team404;

import lombok.Getter;
import org.apache.commons.lang3.tuple.Pair;
import org.bukkit.Material;

import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

@Getter
public class PlayerRevivalCache {
    private final Map<String, List<Pair<Integer, Material>>> respawnablePlayers = new HashMap<>();
    private final Set<String> playersToSpawn = new HashSet<>();

    private static PlayerRevivalCache instance;

    private PlayerRevivalCache() {}

    public static PlayerRevivalCache getInstance() {
        if (instance == null) {
            instance = new PlayerRevivalCache();
        }
        return instance;
    }

    public void addAllRespawnablePlayers(Map<String, List<Pair<Integer, Material>>> players) {
        respawnablePlayers.putAll(players);
    }

    public void addAllPlayersToSpawn(Collection<String> players) {
        playersToSpawn.addAll(players);
    }
}

File: src/main/kotlin/team404/service/PlayerRevivalCache.kt
package team404.service

import org.bukkit.Material

class PlayerRevivalCache private constructor() {
    val respawnablePlayers: MutableMap<String, Map<Material, Int>> = mutableMapOf()
    val playersToSpawn: MutableSet<String> = mutableSetOf()

    fun addAllRespawnablePlayers(players: Map<String, Map<Material, Int>>) {
        players.forEach { (key, value) ->
            respawnablePlayers[key] = value.toMutableMap()
        }
    }

    fun addAllPlayersToSpawn(players: Collection<String>) {
        playersToSpawn.addAll(players)
    }

    companion object {
        @Volatile
        private var instance: PlayerRevivalCache? = null

        fun getInstance(): PlayerRevivalCache =
            instance ?: synchronized(this) {
                instance ?: PlayerRevivalCache().also { instance = it }
            }
    }
}
